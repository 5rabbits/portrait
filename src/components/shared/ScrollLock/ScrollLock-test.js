import { React, toJSON, mount } from 'test/helper'
import ScrollLock from 'shared/ScrollLock'

describe('ScrollLock', () => {
  it('should render with minimum props', () => {
    const component = mount(
      <ScrollLock>
        <div />
      </ScrollLock>,
    )

    expect(toJSON(component)).toMatchSnapshot()
  })

  describe('componentDidMount()', () => {
    it('should invoke toggleLock passing true', () => {
      const component = mount(
        <ScrollLock>
          <div />
        </ScrollLock>,
      )
      const instance = component.instance()

      instance.toggleLock = jest.fn()
      instance.componentDidMount()

      expect(instance.toggleLock).toHaveBeenCalledTimes(1)
      expect(instance.toggleLock).toHaveBeenCalledWith(true)
    })
  })

  describe('componentWillUpdate()', () => {
    it('should invoke toggleLock passing false', () => {
      const component = mount(
        <ScrollLock>
          <div />
        </ScrollLock>,
      )
      const instance = component.instance()

      instance.toggleLock = jest.fn()
      instance.componentWillUpdate()

      expect(instance.toggleLock).toHaveBeenCalledTimes(1)
      expect(instance.toggleLock).toHaveBeenCalledWith(false)
    })
  })

  describe('componentDidUpdate()', () => {
    it('should invoke toggleLock passing true', () => {
      const component = mount(
        <ScrollLock>
          <div />
        </ScrollLock>,
      )
      const instance = component.instance()

      instance.toggleLock = jest.fn()
      instance.componentDidUpdate()

      expect(instance.toggleLock).toHaveBeenCalledTimes(1)
      expect(instance.toggleLock).toHaveBeenCalledWith(true)
    })
  })

  describe('componentWillUnmount()', () => {
    it('should invoke toggleLock passing false', () => {
      const component = mount(
        <ScrollLock>
          <div />
        </ScrollLock>,
      )
      const instance = component.instance()

      instance.toggleLock = jest.fn()
      instance.componentWillUnmount()

      expect(instance.toggleLock).toHaveBeenCalledTimes(1)
      expect(instance.toggleLock).toHaveBeenCalledWith(false)
    })
  })

  describe('onScrollHandler(event)', () => {
    describe('if the element does not has scroll', () => {
      it('should return true', () => {
        const component = mount(
          <ScrollLock>
            <div>
              <div id="scrollable-contents" />
            </div>
          </ScrollLock>,
        )
        const instance = component.instance()

        instance.getNodeMeasures = jest.fn(() => ({
          scrollTop: 0,
          scrollHeight: 10,
          clientHeight: 10,
        }))

        const event = {
          target: component.find('#scrollable-contents').getDOMNode(),
        }
        const returnValue = instance.onScrollHandler(event)

        expect(returnValue).toBe(true)
      })
    })

    describe('if the element has scroll', () => {
      describe('if the user is scrolling down and the scroll is already at the bottom', () => {
        it('should keep the scroll at the bottom', () => {
          const component = mount(
            <ScrollLock>
              <div>
                <div id="scrollable-contents" />
              </div>
            </ScrollLock>,
          )
          const instance = component.instance()
          const node = component.getDOMNode()
          const event = {
            deltaY: 1,
            target: component.find('#scrollable-contents').getDOMNode(),
          }

          instance.getNodeMeasures = jest.fn(() => ({
            scrollTop: 12,
            scrollHeight: 12,
            clientHeight: 10,
          }))

          instance.cancelScrollEvent = jest.fn()
          instance.onScrollHandler(event)

          expect(node.scrollTop).toBe(12)
        })

        it('should invoke cancelScrollEvent passing the event', () => {
          const component = mount(
            <ScrollLock>
              <div>
                <div id="scrollable-contents" />
              </div>
            </ScrollLock>,
          )
          const instance = component.instance()
          const event = {
            deltaY: 1,
            target: component.find('#scrollable-contents').getDOMNode(),
          }

          instance.getNodeMeasures = jest.fn(() => ({
            scrollTop: 12,
            scrollHeight: 12,
            clientHeight: 10,
          }))

          instance.cancelScrollEvent = jest.fn()
          instance.onScrollHandler(event)

          expect(instance.cancelScrollEvent).toHaveBeenCalledTimes(1)
          expect(instance.cancelScrollEvent).toHaveBeenCalledWith(event)
        })
      })

      describe('if the user is scrolling up and the scroll is already at the top', () => {
        it('should keep the scroll at the top', () => {
          const component = mount(
            <ScrollLock>
              <div>
                <div id="scrollable-contents" />
              </div>
            </ScrollLock>,
          )
          const instance = component.instance()
          const node = component.getDOMNode()
          const event = {
            deltaY: -1,
            target: component.find('#scrollable-contents').getDOMNode(),
          }

          instance.getNodeMeasures = jest.fn(() => ({
            scrollTop: 0,
            scrollHeight: 12,
            clientHeight: 10,
          }))

          instance.cancelScrollEvent = jest.fn()
          instance.onScrollHandler(event)

          expect(node.scrollTop).toBe(0)
        })

        it('should invoke cancelScrollEvent passing the event', () => {
          const component = mount(
            <ScrollLock>
              <div>
                <div id="scrollable-contents" />
              </div>
            </ScrollLock>,
          )
          const instance = component.instance()
          const event = {
            deltaY: -1,
            target: component.find('#scrollable-contents').getDOMNode(),
          }

          instance.getNodeMeasures = jest.fn(() => ({
            scrollTop: 0,
            scrollHeight: 12,
            clientHeight: 10,
          }))

          instance.cancelScrollEvent = jest.fn()
          instance.onScrollHandler(event)

          expect(instance.cancelScrollEvent).toHaveBeenCalledTimes(1)
          expect(instance.cancelScrollEvent).toHaveBeenCalledWith(event)
        })
      })

      describe('if the user is scrolling between the scroll bounds', () => {
        it('should return true', () => {
          const component = mount(
            <ScrollLock>
              <div>
                <div id="scrollable-contents" />
              </div>
            </ScrollLock>,
          )
          const instance = component.instance()
          const event = {
            deltaY: 1,
            target: component.find('#scrollable-contents').getDOMNode(),
          }

          instance.getNodeMeasures = jest.fn(() => ({
            scrollTop: 0,
            scrollHeight: 12,
            clientHeight: 10,
          }))

          expect(instance.onScrollHandler(event)).toBe(true)
        })
      })
    })
  })

  describe('cancelScrollEvent(event)', () => {
    it('should call prevent the event propagation in a cross-browser way', () => {
      const event = {
        stopImmediatePropagation: jest.fn(),
        preventDefault: jest.fn(),
        returnValue: null,
      }
      const component = mount(
        <ScrollLock>
          <div />
        </ScrollLock>,
      )
      const instance = component.instance()
      const returnValue = instance.cancelScrollEvent(event)

      expect(event.stopImmediatePropagation).toHaveBeenCalledTimes(1)
      expect(event.preventDefault).toHaveBeenCalledTimes(1)
      expect(event.returnValue).toBe(false)
      expect(returnValue).toBe(false)
    })
  })

  describe('toggleLock(lock)', () => {
    describe('if lock is true', () => {
      it('should bind the "wheel" event to onScrollHandler', () => {
        const component = mount(
          <ScrollLock>
            <div />
          </ScrollLock>,
        )
        const instance = component.instance()
        const node = component.getDOMNode()

        node.addEventListener = jest.fn()
        instance.toggleLock(true)

        expect(node.addEventListener).toHaveBeenCalledTimes(1)
        expect(node.addEventListener).toHaveBeenCalledWith('wheel', instance.onScrollHandler, false)
      })
    })

    describe('if lock is false', () => {
      it('should remove the "wheel" bind', () => {
        const component = mount(
          <ScrollLock>
            <div />
          </ScrollLock>,
        )
        const instance = component.instance()
        const node = component.getDOMNode()

        node.removeEventListener = jest.fn()
        instance.toggleLock(false)

        expect(node.removeEventListener).toHaveBeenCalledTimes(1)
        expect(node.removeEventListener).toHaveBeenCalledWith('wheel', instance.onScrollHandler, false)
      })
    })
  })

  describe('getNodeMeasures(node)', () => {
    it('should return the scroll related node measures', () => {
      const component = mount(
        <ScrollLock>
          <div />
        </ScrollLock>,
      )
      const instance = component.instance()
      const node = {
        scrollTop: 1,
        scrollHeight: 12,
        clientHeight: 10,
      }

      expect(instance.getNodeMeasures(node)).toEqual({
        scrollTop: 1,
        scrollHeight: 12,
        clientHeight: 10,
      })
    })
  })
})
