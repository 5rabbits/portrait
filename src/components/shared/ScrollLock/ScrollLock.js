/* eslint-disable react/no-find-dom-node */

import { PureComponent } from 'react'
import PropTypes from 'prop-types'
import ReactDOM from 'react-dom'

/**
 * This component allows to stop the scroll propagation while using the mouse wheel on an
 * element with overflow.
 *
 * **Note:** This behaviour affects the entire child tree.
 */
export default class ScrollLock extends PureComponent {
  static propTypes = {
    /**
     * The element that needs scroll locking.
     */
    children: PropTypes.node.isRequired,
  }

  componentDidMount() {
    this.toggleLock(true)
  }

  componentWillUpdate() {
    this.toggleLock(false)
  }

  componentDidUpdate() {
    this.toggleLock(true)
  }

  componentWillUnmount() {
    this.toggleLock(false)
  }

  onScrollHandler = (event) => {
    const domNode = event.target.parentNode
    const { scrollTop, scrollHeight, clientHeight } = this.getNodeMeasures(domNode)
    const wheelDelta = event.deltaY
    const isDeltaPositive = wheelDelta > 0

    if (scrollHeight === clientHeight) {
      return true
    }

    if (isDeltaPositive && wheelDelta > scrollHeight - clientHeight - scrollTop) {
      domNode.scrollTop = scrollHeight
      return this.cancelScrollEvent(event)
    }
    else if (!isDeltaPositive && -wheelDelta > scrollTop) {
      domNode.scrollTop = 0
      return this.cancelScrollEvent(event)
    }

    return true
  }

  getNodeMeasures = node => ({
    scrollTop: node.scrollTop,
    scrollHeight: node.scrollHeight,
    clientHeight: node.clientHeight,
  })

  cancelScrollEvent = (event) => {
    event.stopImmediatePropagation()
    event.preventDefault()
    event.returnValue = false // eslint-disable-line no-param-reassign

    return false
  }

  toggleLock = lock => {
    const listenerMethod = lock ? 'addEventListener' : 'removeEventListener'
    const domNode = ReactDOM.findDOMNode(this)

    domNode[listenerMethod]('wheel', this.onScrollHandler, false)
  }

  render() {
    return this.props.children
  }
}
