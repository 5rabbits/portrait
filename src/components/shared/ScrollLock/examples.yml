-
  description: >
    Without scroll locking the mouse wheel event propagates to the document.

    This document page must have scroll for this example to work.
  code: >
    <div style={{ height: 100, width: 240, overflow: 'auto' }}>
      <div style={{ height: 120, padding: 8, backgroundColor: '#fff' }}>
        Scroll with the mouse wheel.
      </div>
    </div>

-
  description: With scroll locking the mouse wheel event doesn't propagate to the document.
  code: >
    <ScrollLock>
      <div style={{ height: 100, width: 240, overflow: 'auto' }}>
        <div style={{ height: 120, padding: 8, backgroundColor: '#fff' }}>
          Scroll with the mouse wheel.
        </div>
      </div>
    </ScrollLock>

-
  description: Independent locking on nested scrollable elements.
  code: >
    <ScrollLock>
      <div
        style={{ height: 120, width: 400, marginTop: 8, overflow: 'auto' }}
        >
        <div style={{ height: 140, padding: 8, backgroundColor: '#fff' }}>
          Scroll on the container.

          <div
            id="scrollable"
            style={{ height: 80, width: 200, margin: '8px auto 0 auto', overflow: 'auto' }}
            >
            <div style={{ height: 100, padding: 8, backgroundColor: '#f00' }}>
              You can scroll on this element too.
            </div>
          </div>
        </div>
      </div>
    </ScrollLock>

