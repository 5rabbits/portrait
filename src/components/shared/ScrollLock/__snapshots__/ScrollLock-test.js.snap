// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ScrollLock should render with minimum props 1`] = `
<ScrollLock>
  <div />
</ScrollLock>
`;
